### ISSUE TITLE
Add support for authentication proxy headers

### ISSUE THREAD

**Issue by cyruz-git** [hoarder-app/hoarder#511]
Hi,

loving this app so far, but it misses a way to authenticate through a proxy (the proxy authenticates the user and forward a header to the webapp).

For your reference, Linkding implements this feature through the LD_ENABLE_AUTH_PROXY environment variable.

Thanks and keep up with the good work!

**Comment by MohamedBassem** [comment]
Hmmm, I'm not sure I'm a big fan of this (at least the linkding implementation). I don't like this idea of the server blindly trusting some header and logging the user in if it's set.

Instead, what we can do is to allow you to pass custom headers (in the extension and mobile apps) that you can use to authenticate to the proxy with, and if the authentication succeeds, the requests then gets forwarded to Hoarder. What do you think?

**Comment by MohamedBassem** [comment]
Or are you not talking about mobile apps/extension and referring to authenticating the web app itself? If yes, then why not just OIDC?

**Comment by cyruz-git** [comment]
Yes I'm talking about the webapp. The custom headers for extensions and mobile apps are a good idea. They make for a great security layer over the web app itself.

The webapp need flexibility though and OIDC is not always the answer, specially if there is a need to maintain consistency over all the app hosted through a proxy.

In my case, for example, i like to enforce a basic auth with a 2FA. Another alternative, would be to authenticate through a JWT, if you think that a plan header is too insecure.

Consider anyway, that a custom header would be insecure only if easy to guess. This is a pretty secure way to login, specially if a trusted proxy is configured to be the only one forwarding that particular header.

**Comment by georgzoeller** [comment]
I want to add my vote for this (auth header). 

It's super useful when you're terminating your entire home network with a caddy that can enforce the headers (e.g. there's no other way to access the app). It's certainly an advanced feature, but it makes self hosting much much easier by allowing the proxy to add as SSO to everything. Grafana supports it and it's pretty good.

**Comment by redge76** [comment]
Yes custom header are one of the easyest solution to secure a selfhosted hoader instance.
It's the standard cloudflare auth https://developers.cloudflare.com/cloudflare-one/identity/service-tokens/ for services.

**Comment by costa-alex** [comment]
+1 for custom headers.

The paperless-ngx app also supports Cloudlfare custom headers and it works pretty good.

**Comment by alexvcoff** [comment]
+1 for custom headers is mobile apps. 
Hoarder instance is behind CloudFlare Access, so currently there’s no way of using in outside home network from iOS devices.

Love the app, but not going to expose it without proper authentication.

**Comment by cyruz-git** [comment]
Yes this is such a crucial feature that it's sad not seeing it in the roadmap. I'm selfhosting more apps than I was when I raised this issue and it's a challenge to create a coherent environment when it comes to authentication. I'm not sure why this issue isn't getting more attention by the community, considering that it would ease the integration in many simple setups.

**Comment by MohamedBassem** [comment]
Just to be clear, what this issue is requesting is different than what the rest of the comments are expecting. This issue is asking for karakeep to support something like what Linkding does by blindly trusting a header coming from the auth proxy as the "logged in" username. At least that's how I understand the original issue. This is the one that I don't agree with.

What the comment is asking for is the ability to pass custom headers in the extension and the mobile app to authenticate to things like cloudflare. That one I 100% agree with and I'm going to implement.

I'm going to repurpose this issue for the latter.

**Comment by MohamedBassem** [comment]
Actually to avoid hijacking this issue, I've filed a new one: #1287. This one I've added to the icebox until there's enough traction for it.

**Comment by georgzoeller** [comment]
For context 

https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/

It's perfectly save if the proxy is configured correctly.

**Comment by caesay** [comment]
I would like this too. Nothing can reach my apps without going through my proxy, and I'd prefer all my auth to happen there as it's far more secure than every app having it's own auth. I've already validated the user and that they have access to this application before they ever reach karakeep. 

Having to configure OIDC separately is added headache for me (to configure) and for my users, as it means they have to effectively authenticate twice to access karakeep. 

Honestly this is my # 1 issue with self hosted apps. 80% of them do support header auth properly, which is amazing, but the other 20% are just a real headache to host. I would even submit PR's to the ones I like (like karakeep) to add header auth, but based on previous comments it doesn't seem like you would accept such a thing. 

Since karakeep has dedicated mobile apps, if this feature was to be implemented, it would be a little more complicated than just a web-only app. The mobile apps would need to support detection of the redirect to the proxy auth and showing that in a webview - as they will not be able to reach the karakeep api before authentication.

**Comment by brettp** [comment]
> This issue is asking for karakeep to support something like what Linkding does by blindly trusting a header coming from the auth proxy as the "logged in" username. At least that's how I understand the original issue. This is the one that I don't agree with.

@MohamedBassem - Could you say more about why you disagree with this?

Header-based authentication is a well-established mechanism that's supported by many (all?) major enterprise IAM providers. Off the top of my head: Okta, Microsoft, and Oracle all support this, with both bare headers and JWTs.

It _does_ require an authenticating proxy, however; that's where the trust lives. The feature would require explicit configuration to enable, so would only affect those who turn it on. Everyone else can ignore it and there would be no change in functionality.

---

> Since karakeep has dedicated mobile apps, if this feature was to be implemented, it would be a little more complicated than just a web-only app. The mobile apps would need to support detection of the redirect to the proxy auth and showing that in a webview - as they will not be able to reach the karakeep api before authentication.

The mobile app supports token-based authentication. It should be reasonably easy to configure the auth proxy to pass bearer tokens directly to the server to authenticate instead of using the proxy's mechanism.

**Comment by MCCC4** [comment]
@MohamedBassem - I had a similar concern about this. In one app I run behind such a proxy I added a check for a pre-shared secret between the proxy and the app. If the app doesn't see it, it refuses to authenticate. The secret is added as a custom header by the proxy. It's a non standard solution but good enough IMO - dead simple to set up (if someone is already configuring auth forwarding this is not a lot of extra work) and misconfiguration results in a safe failure mode.

Alternatively, client-side certs could also be used for this (client being the proxy here) - configure the client cert on the proxy for Karakeep, configure Karakeep to require that cert for auth forwarding.

**Comment by ghost** [comment]
Hi, is it possible that Header-auth is working on the IOS app, but not in the chrome extension?
